### ISSUE TITLE
general performance issue

### ISSUE THREAD

**Issue by convolvatron** [jssmith/ssqlite#11]
optimizations we are consider are
  1) pipeline segmented transfers
  2) connection bonding
  3) iovecs to reduce copying
  4) caching filehandles

note that (2) might obviate (1). (1) may in fact be necessary for correctness if the server evaluates out of order

**Comment by convolvatron** [comment]
5. inserts seem to involve a lot of small writes, probably btree pointers, lsns, etc. it would probably help to batch these between sync(). that may also lower the chances of corrupting the database on fault (quasi-atomic updates)

**Comment by convolvatron** [comment]
(6) up the transfer chunking size

**Comment by jssmith** [comment]
Ok, good learnings + thoughts. Comments:

> 5. inserts seem to involve a lot of small writes...

Just to confirm I that I understand this: you expect we can get efficiencies from batching at the NFS protocol layer. I believe we already have the Nagle algorithm in TCP layer (though now that I think about it I'm not sure whether we want that - maybe we just do all of the batching ourselves).

> 6. up the transfer chunking size

Ah, yes, we should probably experiment with this. We definitely need to renumber these because this seems like low-hanging fruit.

I'll want to make sure that we have all of these changes flagged so that we can measure and compare performance under different approaches.

**Comment by convolvatron** [comment]
right. just like nagle. nagle though is probably going to be a little blind. in particular it flushes with a timer and not with an explicit sync.and i think(?) the size threshold is lower than the tcp mss. sure wrp flags. dynamic api is preferable to compile time.

**Comment by convolvatron** [comment]
oh right, nagle in this case wont help and in fact might be hurting because we block waiting for the remote response on each write

7) turn on NO_DELAY

**Comment by convolvatron** [comment]
note that pipelining segmented transfers might not get us anything, since i'm pretty sure sqlite will max out at a page size. if thats the <= segment size then it kind of doesn't matter (for this application)
